### ISSUE TITLE
Horizontal scaling support on (Meteor) Galaxy Server

### ISSUE THREAD

**Issue by deligence** [RocketChat/meteor-streamer#33]
Hello,

We are working on a project (just like an chatting app) based on meteor and hosted on Galaxy server, and using "rocketchat:streamer" to establish communication b/w multiple devices.

It works fine if the app is deployed on single container but if we scale the app horizontally i.e. if we create more than one container it breaks the functionality, a device connected to one container can't communicate with device connected to other container.

We are not using database for any reason, just emits events with a uniqe string and perform tasks
if a perticular events matches,

Following is code...

const streamer = new Meteor.Streamer('streamer_name');
streamer.emit("unique_pin", {
  event: 'anyEventName',
  ...
});

streamer.on("unique_pin", (data) => {
    const { event } = data;
    if (event === 'anyEventName') {
    	// Do some task if event matches
    }
}

Thanks

**Comment by maxnowack** [comment]
take a look at #26 😉

**Comment by deligence** [comment]
Thanks for your reply

We tried this but the problem is not resolved, there is no change.
But It creates db which I think it is not being used.

import { Meteor } from 'meteor/meteor';
import sync from 'meteor-sync-eventemitter';

@sync({ collectionName: 'stream'})
class SyncedStream extends Meteor.Streamer {
  constructor() {
    super('game');
  }
}
const streamer = new SyncedStream();
streamer.emit(pin, {
      event: 'userJoin'
    });

It works only for single container, are we missing something?

**Comment by maxnowack** [comment]
You haven't configured `meteor-sync-eventemitter` for `rocketchat:streamer`.
It should work if you change the decorator to
```js
@sync({ collectionName: 'stream', emitFn: '_emit' })
```
Take a look at https://github.com/maxnowack/meteor-sync-eventemitter#rocketchatstreamer

**Comment by sampaiodiego** [comment]
Unfortunately this package doesn't scale horizontally automatically. On Rocket.Chat we've done a ["stream matrix" manually](https://github.com/RocketChat/Rocket.Chat/blob/a4ce81203c0c66d6712b60b828403f43ad185bd8/server/stream/streamBroadcast.js#L47), which uses a database to list all running servers and then connect each of them to `emit` changes.

**Comment by evolross** [comment]
It’s so crazy that there still doesn’t seem to be an easy way to connect DDP server to server in a multi-container environment (especially Galaxy). 

Resorting to a database sync seems so crude. We use redos-oplog vent. Which basically puts redis as the sync versus Mongo. Was hunting for ways to remove this round trip and just go straight from server to server.

**Comment by deligence** [comment]
Hi @evolross,

We had tried many different things and finally we have replaced the streamer with PubNub https://www.pubnub.com for our client's app https://www.playfactile.com

**Comment by evolross** [comment]
Right... that's basically what Redis does for us. It's a 3rd party server they all sync with. Wish it could be more direct.

I'm going to experiment with somehow figuring out the Galaxy URL and param programatticaly and see if `DDP.connect` works with that to _directly_ connect to a container. Then it will just be a matter of keeping the container URLs in the database and heartbeat'ing them to make sure they're still alive and update when one gets replaced, crashes, etc.  That just seems cleaner, doesn't put _another_ server/service in the middle, and should reduce lag.
